node/src/components/consensus/era_supervisor.rs

```diff
@@ -298,124 +286,210 @@ where
-            new_faulty,
-            faulty,
-            validators,
-        );
-
-        // Mark validators as faulty for which we have evidence in a recent era.
-        for e_id in self.iter_past_other(era_id, self.bonded_eras()) {
-            if let Some(old_era) = self.active_eras.get_mut(&e_id) {
-                for pub_key in old_era.consensus.validators_with_evidence() {
-                    let proposed_blocks = era.resolve_evidence_and_mark_faulty(pub_key);
-                    if !proposed_blocks.is_empty() {
-                        error!(
-                            ?proposed_blocks,
-                            era = e_id.value(),
-                            "unexpected block in new era"
-                        );
-                    }
-                }
+        // Mark validators as faulty for which we have evidence in the previous era.
+        for pub_key in validators_with_evidence {
+            let proposed_blocks = self
+                .era_mut(era_id)
+                .resolve_evidence_and_mark_faulty(&pub_key);
+            if !proposed_blocks.is_empty() {
+                error!(
+                    ?proposed_blocks,
+                    era = era_id.value(),
+                    "unexpected block in new era"
+                );
             }
         }
 
-        let _ = self.active_eras.insert(era_id, era);
-        let oldest_bonded_era_id = oldest_bonded_era(&self.protocol_config, era_id);
-        // Clear the obsolete data from the era whose validators are unbonded now. We only retain
-        // the information necessary to validate evidence that units in still-bonded eras may refer
-        // to for cross-era fault tracking.
-        if let Some(evidence_only_era_id) = oldest_bonded_era_id.checked_sub(1) {
-            trace!(era = evidence_only_era_id.value(), "clearing unbonded era");
+        // Clear the obsolete data from the era before the previous one. We only retain the
+        // information necessary to validate evidence that units in the two most recent eras may
+        // refer to for cross-era fault tracking.
+        if let Some(evidence_only_era_id) = self.current_era.checked_sub(PAST_ACTIVE_ERAS) {
             if let Some(era) = self.active_eras.get_mut(&evidence_only_era_id) {
+                trace!(era = evidence_only_era_id.value(), "clearing unbonded era");
                 era.consensus.set_evidence_only();
             }
         }
-        // Remove the era that has become obsolete now: The oldest bonded era could still receive
-        // units that refer to evidence from any era that was bonded when it was the current one.
-        let oldest_evidence_era_id = oldest_bonded_era(&self.protocol_config, oldest_bonded_era_id);
-        if let Some(obsolete_era_id) = oldest_evidence_era_id.checked_sub(1) {
+
+        // Remove the era that has become obsolete now: We keep only three in memory.
+        if let Some(obsolete_era_id) = self.current_era.checked_sub(3) {
```

Comment:
Can we use `PAST_ACTIVE_ERAS + 1` here, instead of `3`?

Reply:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/f417ff94706385eda1d3d134cadcc7fec3f6dfd0.